File: coolrmi/hu.qgears.coolrmi/src/hu/qgears/coolrmi/serializer/PortableSerializer.java
package hu.qgears.coolrmi.serializer;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import hu.qgears.coolrmi.remoter.AbstractSerializer;
import hu.qgears.coolrmi.remoter.IReplaceSerializable;

public class PortableSerializer extends AbstractSerializer {
	private static final TypeSerializer[] serializers = new TypeSerializer[] {
		/*Null*/  new NullSerializer(),
		/*Bool*/  new BooleanSerializer(),
		/*Int8*/  new ByteSerializer(),
		/*Int16*/ new ShortSerializer(),
		/*Int32*/ new IntSerializer(),
		/*Int64*/ new LongSerializer(),
		/*Char*/  new CharSerializer(),
		/*Float*/ new FloatSerializer(),
		/*Double*/new DoubleSerializer(),
		/*Enum*/  new EnumSerializer(),
		/*String*/new StringSerializer(),
		/*Object*/new ObjectSerializer(),
		/*Array*/ new ArraySerializer(),
		/*Exception*/ new ExceptionSerializer(),
		/*List*/  new ListSerializer(),
	};

	private static final Map<Class<?>, TypeSerializer> classMap =
			new HashMap<Class<?>, TypeSerializer>();
	private static final List<TypeSerializer> specialSerializer =
			new ArrayList<TypeSerializer>();

	private Map<String, String> javaToPortableNameMap = new HashMap<String, String>();
	private Map<String, String> portableToJavaNameMap = new HashMap<String, String>();

	private Map<String, String> javaToPortableNamespaceMap = new HashMap<String, String>();
	private Map<String, String> portableToJavaNamespaceMap = new HashMap<String, String>();

	public void addMapping(String java, String portable) {
		javaToPortableNameMap.put(java, portable);
		portableToJavaNameMap.put(portable, java);
	}

	public void addNamespaceMapping(String ns, String portable) {
		javaToPortableNamespaceMap.put(ns, portable);
		portableToJavaNamespaceMap.put(portable, ns);
	}

	static {
		for (int i = 0; i < serializers.length; ++i) {
			TypeSerializer s = serializers[i];
			if (s != null) {
				if (s.type.ordinal() != i) {
					throw new AssertionError("Invalid serializers");
				}

				if (s.serializedClass != null) {
					classMap.put(s.serializedClass, s);
				}
				if (s.primitiveClass != null) {
					classMap.put(s.primitiveClass, s);
				}
				if (s.canSerializeIsSpecial()) {
					specialSerializer.add(s);
				}
			}
		}
	}

	public PortableSerializer() {
		// add standard types
		addMapping(Object.class.getName(), "Object");
		addNamespaceMapping("hu.qgears.coolrmi", "coolrmi");
		addNamespaceMapping("hu.qgears.coolrmi.messages", "coolrmi.messages");
	}

	@Override
	public byte[] serialize(Object o) throws IOException {
		ByteArrayOutputStream os = new ByteArrayOutputStream();
		serialize(os, o, null);
		return os.toByteArray();
	}

	private final JavaType replaceSerializableType = new JavaType(
			IReplaceSerializable.class);
	TypeSerializer getSerializer(JavaType typ) {
		if (getServiceRegistry().getReplacer(typ.getCls()) != null) {
			typ = replaceSerializableType;
		}

		TypeSerializer s = classMap.get(typ.getCls());
		if (s == null) {
			for (TypeSerializer sc : specialSerializer) {
				if (sc.canSerialize(this, typ)) {
					return sc;
				}
			}

			throw new RuntimeException(
					"Unserializable class " + typ.getCls().getName());
		} else {
			return s;
		}
	}

	TypeSerializer getSerializer(int idx) {
		return serializers[idx];
	}

	void serialize(OutputStream os, Object o, JavaType typ)
			throws IOException {
		TypeSerializer s;
		o = getServiceRegistry().replaceObject(o);

		boolean full = typ == null;
		if (typ == null && o == null) {
			s = serializers[TypeId.Null.ordinal()];
		} else {
			if (typ == null) {
				typ = new JavaType(o);
			}
			s = getSerializer(typ);
			if (s.isPolymorphic()) {
				full = true;
				if (o != null) {
					typ = new JavaType(typ, o);
					s = getSerializer(typ);
				} else {
					s = serializers[TypeId.Null.ordinal()];
				}
			}
		}

		if (full) {
			s.writeType(this, os, typ);
		}
		s.serialize(this, os, o, typ);
	}

	@Override
	public Object deserialize(byte[] bs) throws Exception {
		ByteArrayInputStream is = new ByteArrayInputStream(bs);
		return deserialize(is, null);
	}

	Object deserialize(InputStream is, JavaType typ) throws Exception {
		JavaType desType;
		TypeSerializer ser;

		if (typ != null) {
			ser = getSerializer(typ);
			if (ser.isPolymorphic()) {
				ser = getSerializer(is.read());
				desType = ser.readType(this, is);
			} else {
				desType = typ;
			}
		} else {
			ser = getSerializer(is.read());
			desType = ser.readType(this, is);
		}

		Object ret = ser.deserialize(this, is, desType);
		if (ret instanceof IReplaceSerializable) {
			return ((IReplaceSerializable) ret).readResolve();
		}
		return ret;
	}

	String getPortableClassName(JavaType typ) {
		String javaName = typ.getCls().getName();
		String portableName = javaToPortableNameMap.get(javaName);
		if (portableName != null) {
			return portableName;
		}

		String ns = typ.getCls().getPackage().getName();
		portableName = javaToPortableNamespaceMap.get(ns);
		if (portableName != null) {
			portableName = portableName + '.'
					+ javaName.substring(javaName.lastIndexOf('.') + 1);
		} else {
			System.err.println("Warning: no portable name for " + javaName);
			portableName = javaName;
		}

		// cache for future
		addMapping(javaName, portableName);
		return portableName;
	}

	public void writeClassName(OutputStream os, JavaType typ) throws IOException {
		Utils.writeString(os, getPortableClassName(typ));

		for (JavaType x : typ.getGenericTypes()) {
			writeClassName(os, x);
		}
	}


	String getJavaClassName(String portableName) {
		String javaName = portableToJavaNameMap.get(portableName);
		if (javaName != null) {
			return javaName;
		}

		int dotPos = portableName.lastIndexOf('.');
		String ns = portableName.substring(0, dotPos);
		javaName = portableToJavaNamespaceMap.get(ns);
		if (javaName != null) {
			javaName = javaName + '.' + portableName.substring(dotPos + 1);
		} else {
			javaName = portableName;
		}

		// cache for future
		addMapping(javaName, portableName);
		return javaName;
	}

	public JavaType readClassName(InputStream is)
			throws ClassNotFoundException, IOException {
		Class<?> cls = loadClass(Utils.readString(is));

		Type[] generics = cls.getTypeParameters();
		JavaType[] jGenerics = new JavaType[generics.length];
		for (int i = 0; i < generics.length; ++i) {
			jGenerics[i] = readClassName(is);
		}
		return new JavaType(cls, jGenerics);
	}

	private Class<?> loadClass(String name) throws ClassNotFoundException {
		return getClassLoader().loadClass(getJavaClassName(name));
	}
}

File: coolrmi/CoolRMI.Net/Serializer/PortableSerializer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using CoolRMI.Net.Remoter;

namespace CoolRMI.Net.Serializer
{
    public class PortableSerializer : AbstractSerializer
    {
        private static readonly TypeSerializer[] Serializers = {
            /*Null*/  new NullSerializer(),
            /*Bool*/  new BoolSerializer(),
            /*Int8*/  new ByteSerializer(),
            /*Int16*/ new ShortSerializer(),
            /*Int32*/ new IntSerializer(),
            /*Int64*/ new LongSerializer(),
            /*Char*/  new CharSerializer(),
            /*Float*/ new FloatSerializer(),
            /*Double*/new DoubleSerializer(),
            /*Enum*/  new EnumSerializer(),
            /*String*/new StringSerializer(),
            /*Object*/new ObjectSerializer(),
            /*Array*/ new ArraySerializer(),
            /*Exception*/ new ExceptionSerializer(),
            /*List*/  new ListSerializer(),
        };

        private static readonly Dictionary<Type, TypeSerializer> ClassMap =
            new Dictionary<Type, TypeSerializer>();
        private static readonly List<TypeSerializer> SpecialSerializers =
            new List<TypeSerializer>();

        private readonly Dictionary<string, string> dotNetToPortableNameMap =
            new Dictionary<string, string>();
        private readonly Dictionary<string, string> portableToDotNetNameMap =
            new Dictionary<string, string>();

        private struct NamespaceEntry : IEquatable<NamespaceEntry>
        {
            public readonly string Namespace;
            public readonly AssemblyName AssemblyName;

            public NamespaceEntry(string ns, AssemblyName assemblyName)
            {
                Namespace = ns;
                AssemblyName = assemblyName;
            }

            public bool Equals(NamespaceEntry other)
            {
                return string.Equals(Namespace, other.Namespace) &&
                       Equals(AssemblyName.FullName,
                           other.AssemblyName.FullName);
            }

            public override bool Equals(object obj)
            {
                if (ReferenceEquals(null, obj)) return false;
                return obj is NamespaceEntry && Equals((NamespaceEntry) obj);
            }

            public override int GetHashCode()
            {
                unchecked
                {
                    return (Namespace.GetHashCode() * 397) ^
                           AssemblyName.FullName.GetHashCode();
                }
            }
        }

        private readonly Dictionary<NamespaceEntry, string>
            dotNetToPortableNamespaceMap =
                new Dictionary<NamespaceEntry, string>();
        private readonly Dictionary<string, NamespaceEntry>
            portableToDotNetNamespaceMap =
                new Dictionary<string, NamespaceEntry>();

        #region Replacer

        private readonly Dictionary<Type, CoolRMIReplaceEntry> replaceTypes =
            new Dictionary<Type, CoolRMIReplaceEntry>();

        public void AddReplaceType(CoolRMIReplaceEntry entry)
        {
            replaceTypes.Add(entry.TypeToReplace, entry);
        }

        public object ReplaceObject(object obj)
        {
            if (obj == null) return null;

            var replacer = GetReplacer(obj.GetType());
            return replacer == null ? obj : replacer.DoReplace(obj);
        }

        public CoolRMIReplaceEntry GetReplacer(Type type)
        {
            CoolRMIReplaceEntry found;
            if (replaceTypes.TryGetValue(type, out found)) return found;

            foreach (var e in replaceTypes.Values)
            {
                if (e != null && e.TypeToReplace.IsAssignableFrom(type))
                {
                    found = e;
                    break;
                }
            }

            replaceTypes.Add(type, found);
            return found;
        }

        #endregion

        public void AddMapping(string dotNet, string portable)
        {
            dotNetToPortableNameMap.Add(dotNet, portable);
            portableToDotNetNameMap.Add(portable, dotNet);
        }

        public void AddNamespaceMapping(string ns, AssemblyName assemblyName,
            string portable)
        {
            dotNetToPortableNamespaceMap.Add(
                new NamespaceEntry(ns, assemblyName), portable);
            portableToDotNetNamespaceMap.Add(portable,
                new NamespaceEntry(ns, assemblyName));
        }

        static PortableSerializer()
        {
            for (int i = 0; i < Serializers.Length; ++i)
            {
                var s = Serializers[i];
                if (s == null) continue;

                if ((int) s.Type != i)
                    throw new InvalidProgramException("Invalid serializers");

                if (s.DotNetType != null) ClassMap.Add(s.DotNetType, s);
                if (s.CanSerializeIsSpecial) SpecialSerializers.Add(s);
            }
        }

        public PortableSerializer()
        {
            AddMapping(typeof(object).AssemblyQualifiedName, "Object");
            AddNamespaceMapping("CoolRMI.Net", GetType().Assembly.GetName(), "coolrmi");
            AddNamespaceMapping("CoolRMI.Net.Messages", GetType().Assembly.GetName(), "coolrmi.messages");
        }

        public override byte[] Serialize(object o)
        {
            var memstream = new MemoryStream();
            var bw = new BinaryWriter(memstream);
            Serialize(bw, o, null);
            bw.Flush();
            return memstream.GetBuffer();
        }

        internal TypeSerializer GetSerializer(Type typ)
        {
            if (GetReplacer(typ) != null) typ = typeof(IReplaceSerializable);

            TypeSerializer s;
            if (ClassMap.TryGetValue(typ, out s)) return s;

            foreach (var sc in SpecialSerializers)
                if (sc.CanSerialize(this, typ))
                    return sc;
            throw new Exception("Unserializable class " + typ);
        }

        internal TypeSerializer GetSerializer(int idx)
        {
            return Serializers[idx];
        }

        internal void Serialize(BinaryWriter bw, object o, Type typ)
        {
            TypeSerializer s;
            o = ReplaceObject(o);

            bool full = typ == null;
            if (typ == null && o == null)
            {
                s = Serializers[(int) TypeId.Null];
            }
            else
            {
                if (typ == null) typ = o.GetType();
                s = GetSerializer(typ);
                if (s.IsPolymorphic)
                {
                    full = true;
                    if (o != null)
                    {
                        typ = o.GetType();
                        s = GetSerializer(typ);
                    }
                    else
                    {
                        s = Serializers[(int) TypeId.Null];
                    }
                }
            }

            if (full) s.WriteType(this, bw, typ);
            s.Serialize(this, bw, o, typ);
        }

        public override object Deserialize(byte[] data)
        {
            var memstream = new MemoryStream(data);
            return Deserialize(new BinaryReader(memstream), null);
        }

        internal object Deserialize(BinaryReader br, Type typ)
        {
            Type desType;
            TypeSerializer ser;

            if (typ != null)
            {
                ser = GetSerializer(typ);
                if (ser.IsPolymorphic)
                {
                    ser = GetSerializer(br.ReadByte());
                    desType = ser.ReadType(this, br);
                }
                else
                {
                    desType = typ;
                }
            }
            else
            {
                ser = GetSerializer(br.ReadByte());
                desType = ser.ReadType(this, br);
            }

            var ret = ser.Deserialize(this, br, desType);
            var serializable = ret as IReplaceSerializable;
            return serializable != null ? serializable.ReadResolve() : ret;
        }

        internal string GetPortableClassName(Type typ)
        {
            if (typ.IsGenericType) typ = typ.GetGenericTypeDefinition();

            var dotNetName = typ.AssemblyQualifiedName;
            string portableName;
            if (dotNetToPortableNameMap.TryGetValue(dotNetName, out portableName))
                return portableName;

            if (dotNetToPortableNamespaceMap.TryGetValue(
                new NamespaceEntry(typ.Namespace, typ.Assembly.GetName()),
                out portableName))
            {
                portableName = portableName + "." + typ.Name;
            }
            else
            {
                Console.WriteLine("Warning: no portable name for " + dotNetName);
                portableName = dotNetName;
            }

            // cache for future
            AddMapping(dotNetName, portableName);
            return portableName;
        }

        public void WriteClassName(BinaryWriter bw, Type typ)
        {
            bw.WritePString(GetPortableClassName(typ));

            foreach (var t in typ.GenericTypeArguments)
                WriteClassName(bw, t);
        }

        internal string GetDotNetClassName(string portableName)
        {
            string dotNetName;
            if (portableToDotNetNameMap.TryGetValue(portableName, out dotNetName))
                return dotNetName;

            var dotPos = portableName.LastIndexOf('.');
            var ns = portableName.Substring(0, dotPos);
            NamespaceEntry entry;
            if (portableToDotNetNamespaceMap.TryGetValue(ns, out entry))
            {
                var name = portableName.Substring(dotPos + 1);
                dotNetName = entry.Namespace + '.' + name + ", " +
                             entry.AssemblyName;
            }
            else
            {
                Console.WriteLine("Warning: can't resolve portable name " + portableName);
                dotNetName = portableName;
            }

            // cache for future
            AddMapping(dotNetName, portableName);
            return dotNetName;
        }

        public Type ReadClassName(BinaryReader rd)
        {
            var typ = Type.GetType(GetDotNetClassName(rd.ReadPString()));
            if (typ != null && typ.IsGenericType)
            {
                var args = typ.GetGenericArguments();
                for (int i = 0; i < args.Length; ++i)
                    args[i] = ReadClassName(rd);
                typ = typ.MakeGenericType(args);
            }
            return typ;
        }
    }
}
